src/backend/distributed/master/master_stage_protocol.c

```diff
@@ -556,41 +528,35 @@ CreateShardsOnWorkers(Oid distributedRelationId, List *shardPlacements,
 			partitionRelationShard->shardId = shardId;
 
 			relationShardList = list_make2(parentRelationShard, partitionRelationShard);
-			placementAccessList = BuildPlacementDDLList(shardPlacement->groupId,
-														relationShardList);
-
-			connection = GetPlacementListConnection(connectionFlags, placementAccessList,
-													NULL);
-		}
-		else
-		{
-			connection = GetPlacementConnection(connectionFlags, shardPlacement,
-												NULL);
 		}
 
-		if (useExclusiveConnection)
+		commands = WorkerCreateShardCommandList(distributedRelationId, shardIndex,
+												shardId, ddlCommandList,
+												foreignConstraintCommandList);
+		foreach(commandCell, commands)
 		{
-			ClaimConnectionExclusively(connection);
-			claimedConnectionList = lappend(claimedConnectionList, connection);
+			const char *command = lfirst(commandCell);
```

Comment:
Hmm, I'm still thinking on this. We probably never have other places with multiple commands in the `task->queryString`.  I'm trying to understand the implications of this.

I think it is also not a good idea to make each command as a separate task because there is no guarantee of execution of the tasks. So we might try to create fkey before the shard is created etc.

We're implicitly always going from 0th to last element in the task list. But still sounds dangerous. 

In sequential execution PR, @marcocitus mentioned about creating dependencies among the tasks. Still, that probably require a big change.

Any other alternatives can we think of?

Reply:
I think concatenating query strings is acceptable and will help improve performance. The TransactionStateMachine will happily loop around in REMOTE_TRANSACTION_CLEARING_RESULTS after sending it. 

There are other cases in the Citus code base where we concatenate commands. The task-tracker executor does it a lot.

Reply:
Thanks for the note! Then I'll stick to this solution

Reply:
We also do this in the unified executor already whenever we send BEGIN.